crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -62,15 +124,27 @@ pub async fn sync(
             Some((number, hash)) => (number + 1, Some(hash)),
             None => (StarknetBlockNumber::GENESIS, None),
         };
-
         let t_block = std::time::Instant::now();
+
         let block = loop {
             match download_block(next, chain, head_hash, &sequencer).await? {
                 DownloadBlock::Block(block) => break block,
                 DownloadBlock::AtHead => {
-                    let poll_interval = head_poll_interval(chain);
-                    tracing::info!(poll_interval=?poll_interval, "At head of chain");
-                    tokio::time::sleep(poll_interval).await;
+                    // Poll pending if it is enabled, otherwise just wait to poll head again.
+                    match pending_poll_interval {
+                        Some(interval) => {
+                            let head_hash = head_hash
+                                .expect("Head hash should exist when entering pending mode");
```

Comment:
Right now `download_block` can return `AtHead` when the genesis block is (still) missing however strange this scenario might sound. Maybe it's worth adding a special variant besides `AtHead` to handle this situation or at least change the `expect` to something like `bail!("failed to download genesis, cannot enter pending mode")`?

Reply:
I agree that there is a missing possibility here, but I think I'd rather not handle that extremely niche case here as well. We need to refactor / redo this for p2p to work, so I'm happy just waiting for that to "fix" it.

Reply:
You're right, this is extremely unlikely (if even possible :thinking: ).